modules/addons/external-dns/main.tf

```diff
@@ -0,0 +1,49 @@
+provider "helm" {
+  kubernetes {
+    host                   = var.k8s_host
+    cluster_ca_certificate = base64decode(var.k8s_cluster_ca_certificate)
+    token                  = var.k8s_client_token
+  }
+}
+
+resource "helm_release" "external_dns" {
+  name             = "external-dns"
+  repository       = "https://charts.bitnami.com/bitnami"
+  chart            = "external-dns"
+  create_namespace = true
+  namespace        = "external-dns"
+  timeout          = 900
+
+  set {
+    name  = "policy"
+    value = "upsert-only"
+  }
+
+  set {
+    name  = "registry"
+    value = "txt"
+  }
+
+  set {
+    name  = "txtOwnerId"
+    value = var.cluster_name
+  }
+
+  # Since we are enabling the Istio Gateway source, it is important to set this filter
+  # so we don't create DNS records for any random or internal hostnames defined in
+  # the TSB IngressGateways.
+  set {
+    name  = "domainFilters"
+    value = "{${var.dns_zone}}"
+  }
+
+  set {
+    name  = "sources"
+    value = "{service,ingress,istio-gateway}"
+  }
+
+  values = [templatefile("${path.module}/manifests/values-${var.dns_provider}.yaml.tmpl", {
```

Comment:
i understand why we are trying to do `dns_provider` here, however this is not necessary cloud agnostic approach here, so let's make a call if we want to have `external-dns` as a single module that accounts sub-zone creation as well service account management with the post-install function after it, instead of trying to do extra lookups that may drive unnecessary complexity, i.e. the code should be easily picked by the field team and customer... if needed

Reply:
There is no added complexity here. As I explained above, you can't parameterize module paths and you'll need a provider-specific module anyway in the add-ons folder, so there is no added complexity.
You have one generic module that requires the provider parameters and the credentials, and you call it from each provider module in the add-ons (that you have to create anyway) just with the provider-specific data. Less code, less modules tor season about.

Reply:
if the dependencies are managed within the module - there is no need even to pass them externally so having "credentials" in outbound way like it is defined now, does create a one more level of execution to account, even though we manage it via `Makefile` but this is not an ideal thing that we want to do, and why it is bad to have all the dependencies under the module that only needs it?  i.e. accounting we follow module/gcp/external-dns  with sub-zone approach.

Reply:
You keep assuming that this can be all done in one module, and we can't.

Credentials must be created _once_, external-dns must be installed N times reusing those credentials created once, so these two cases are better modeled in two different modules. No matter where you place the code, in the add-ons folder or in the generic modules you still need two modules.

Reply:
there is `outputs` folder to store generated artefacts and module function can store and lookup file over there.